segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/metadata/BaseMetadataStore.java

```diff
@@ -0,0 +1,635 @@
+     */
+    @Override
+    public void create(MetadataTransaction txn, StorageMetadata metadata) throws StorageMetadataException {
+        Preconditions.checkArgument(null != txn);
+        Preconditions.checkArgument(null != metadata);
+        Preconditions.checkArgument(null != metadata.getKey());
+        Map<String, TransactionData> txnData = txn.getData();
+        txnData.put(metadata.getKey(), TransactionData.builder()
+                .key(metadata.getKey())
+                .value(metadata)
+                .version(txn.getVersion())
+                .build());
+    }
+
+    /**
+     * Deletes a metadata record given the key.
+     *
+     * @param txn Transaction.
+     * @param key key to use to retrieve metadata.
+     * @throws StorageMetadataException Exception related to storage metadata operations.
+     */
+    @Override
+    public void delete(MetadataTransaction txn, String key) throws StorageMetadataException {
+        Preconditions.checkArgument(null != txn);
+        Preconditions.checkArgument(null != key);
+        Map<String, TransactionData> txnData = txn.getData();
+
+        TransactionData data = TransactionData.builder().key(key).build();
+        TransactionData oldData = txnData.putIfAbsent(key, data);
+        if (null != oldData) {
+            data = oldData;
+        }
+        data.setValue(null);
+        data.setPersisted(false);
+        data.setVersion(txn.getVersion());
+    }
+
+    /**
+     * {@link AutoCloseable#close()} implementation.
+     */
+    @Override
+    public void close() throws Exception {
+        ArrayList<TransactionData> modifiedValues = new ArrayList<>();
+        bufferedTxnData.entrySet().stream().filter(entry -> !entry.getValue().isPersisted() && !entry.getValue().isPinned()).forEach(entry -> modifiedValues.add(entry.getValue()));
+        if (modifiedValues.size() > 0) {
+            writeAll(modifiedValues);
+        }
+    }
+
+    /**
+     * Explicitly marks the store as fenced.
+     * Once marked fenced no modifications to data should be allowed.
+     */
+    public void markFenced() {
+        this.fenced.set(true);
+    }
+
+    /**
+     * Return whether record for the given key is pinned or not.
+     *
+     * @param key Key to check.
+     * @return True if record for the given key is pinned, false otherwise.
+     */
+    protected boolean isPinnedKey(String key) {
+        return pinnedKeys.contains(key);
+    }
+
+    /**
+     * Retrieves the current version number.
+     *
+     * @return current version number.
+     */
+    protected long getVersion() {
+        return version.get();
+    }
+
+    /**
+     * Sets the current version number.
+     *
+     * @param version Version to set.
+     */
+    protected void setVersion(long version) {
+        this.version.set(version);
+    }
+
+    /**
+     * Stores the transaction data.
+     */
+    @Builder(toBuilder = true)
+    @Data
+    public static class TransactionData implements Serializable {
```

Comment:
Is this class thread safe? If more than one thread touches it during its lifetime, even if not concurrently, it must be made thread safe.

Reply:
This class is thread safe. Instance of these objects are not shared between transactions.